lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+                              size_t IndexOffset) {
+    // Qualifiers here are not allowed since this must point to a simple struct.
     revng_assert(StackArgumentTypes.Qualifiers.empty());
+
     auto *Unqualified = StackArgumentTypes.UnqualifiedType.get();
-    if (not Unqualified)
-      return {};
+    if (!Unqualified) {
+      // If there is no type, it means that the importer responsible for this
+      // function didn't detect any stack arguments and avoided creating
+      // a new empty type.
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
 
     auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
     revng_assert(Pointer != nullptr,
                  "`RawFunctionType::StackArgumentsType` must be a struct");
     const model::StructType &Types = *Pointer;
 
+    // If the struct is empty, there's no arguments to extract from it.
+    if (Types.Fields.empty()) {
+      revng_assert(Types.size() == 0);
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
+
+    // Compute the alignment of the first argument.
+    auto CurrentAlignment = *abi::naturalAlignment(Types.Fields.begin()->Type,
+                                                   ABI);
+    if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+      // Abandon cases where alignment is equal to 0 or is not a power of two.
+      return std::nullopt;
+    }
+
+    // Process each of the fields, converting them into arguments separately.
     llvm::SmallVector<model::Argument, 8> Result;
-    for (const model::StructField &Field : Types.Fields) {
+    for (auto Iterator = Types.Fields.begin();
+         Iterator != std::prev(Types.Fields.end());
+         ++Iterator) {
+      const auto &CurrentArgument = *Iterator;
+      auto MaybeSize = CurrentArgument.Type.size();
+      revng_assert(MaybeSize.has_value() && MaybeSize.value() != 0);
+      uint64_t CurrentSize = paddedSizeOnStack(MaybeSize.value());
+
+      const auto &NextArgument = *std::next(Iterator);
+      auto NextAlignment = *abi::naturalAlignment(NextArgument.Type, ABI);
+      if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+        // Abandon cases where alignment is equal to 0 or is not a power of two.
+        return std::nullopt;
+      }
+
+      if (CurrentArgument.Offset + CurrentSize != NextArgument.Offset
+          || (CurrentArgument.Offset + CurrentSize) % NextAlignment != 0) {
+        // Abandon any cases where natural alignment causes a padding "hole"
+        // to be required.
+        //
+        // TODO: we should probably preprocess the input to fill such holes in
+        //       before trying to convert the functions.
+        return std::nullopt;
+      }
+
+      // Create the argument from this field.
       model::Argument &New = Result.emplace_back();
       New.Index = IndexOffset++;
-      New.Type = Field.Type;
-      New.CustomName = Field.CustomName;
-      New.OriginalName = Field.OriginalName;
+      New.Type = CurrentArgument.Type;
```

Comment:
Here we're verbatim forwarding the argument type from `Raw` to `CABI`. Is this safe? Is passing a `uint64_t` the same as passing a `struct { char X[8]; }`?

Reply:
Actually yes, this is a good point! Thanks for pointing it out.

To clarify this is indeed the same for any non-position based ABIs (as long as the struct is naturally-padded), but this would cause problems when properly tested under Microsoft-x64 and such.

Reply:
If you now we don't support something bail out and leave a TODO, half baked support of stuff is evil!

Reply:
(the code in question is gone, but I'm keeping this open as a remainder to double-check this since it's very important to get right)

Reply:
@aleclearmind, about this, it requires unnatural alignment support, which we decided to postpone for now. What should I do? Just close it and leave a todo for the next PR?

Reply:
Make sure the problem is noted somewhere and, if appropriate, leave a `TODO` in the code too.

Reply:
I left a couple of TODOs.